hoot-core/src/main/cpp/hoot/core/io/OsmApiDbAwareHootApiDbWriter.cpp

```diff
@@ -0,0 +1,290 @@
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2016 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+#include "OsmApiDbAwareHootApiDbWriter.h"
+
+// hoot
+#include <hoot/core/Factory.h>
+
+namespace hoot
+{
+
+HOOT_FACTORY_REGISTER(OsmMapWriter, OsmApiDbAwareHootApiDbWriter)
+
+OsmApiDbAwareHootApiDbWriter::OsmApiDbAwareHootApiDbWriter() :
+HootApiDbWriter()
+{
+}
+
+OsmApiDbAwareHootApiDbWriter::~OsmApiDbAwareHootApiDbWriter()
+{
+  _osmApiDb.close();
+}
+
+void OsmApiDbAwareHootApiDbWriter::open(QString urlStr)
+{
+  if (!_osmApiDb.isSupported(ConfigOptions().getHootapiDbWriterOsmapidbUrl()))
+  {
+    throw HootException(
+      "The OsmApiDbAwareHootApiDbWriter requires a valid OSM API DB URL for element id sequencing.");
+  }
+
+  HootApiDbWriter::open(urlStr);
+  _osmApiDb.open(ConfigOptions().getHootapiDbWriterOsmapidbUrl());
+}
+
+long OsmApiDbAwareHootApiDbWriter::_getRemappedElementId(const ElementId& eid)
+{
+  long retVal = -1;
+
+  switch(eid.getType().getEnum())
+  {
+    case ElementType::Node:
+
+      if (_nodeRemap.count(eid.getId()) == 1)
+      {
+        retVal = _nodeRemap.at(eid.getId());
+      }
+      else
+      {
+        retVal = _osmApiDb.getNextId(ElementType::Node);
+        _nodeRemap[eid.getId()] = retVal;
+        if (_outputMappingFile.length() > 0)
+        {
+          _sourceNodeIds.insert(eid.getId());
+        }
+      }
+
+      break;
+
+    case ElementType::Way:
+
+      if (_wayRemap.count(eid.getId()) == 1)
+      {
+        retVal = _wayRemap.at(eid.getId());
+      }
+      else
+      {
+        retVal = _osmApiDb.getNextId(ElementType::Way);
+        _wayRemap[eid.getId()] = retVal;
+        if ( _outputMappingFile.length() > 0 )
+        {
+          _sourceWayIds.insert(eid.getId());
+        }
+      }
+
+      break;
+
+    case ElementType::Relation:
+
+      if (_relationRemap.count(eid.getId()) == 1)
+      {
+        retVal = _relationRemap.at(eid.getId());
+      }
+      else
+      {
+        retVal = _osmApiDb.getNextId(ElementType::Relation);
+        _relationRemap[eid.getId()] = retVal;
+        if ( _outputMappingFile.length() > 0 )
+        {
+          _sourceRelationIds.insert(eid.getId());
+        }
+      }
+
+      break;
+
+  default:
+
+    LOG_ERROR("Tried to create or remap ID for invalid type");
+    throw NotImplementedException();
+
+    break;
+  }
+
+  return retVal;
+}
+
+vector<long> OsmApiDbAwareHootApiDbWriter::_remapNodes(const vector<long>& nids)
+{
+  vector<long> result(nids.size()); // Creates the vector and fills nids.size number of zeroes
+
+  for (size_t i = 0; i < nids.size(); i++)
+  {
+    // This is only called when adding nodes for a way.  If a way has a node we
+    //    did not successfully create a mapping for when importing nodes,
+    //    we can't continue
```

Comment:
Should you just call `_getRemappedElementId`? Simplify the function and it should handle remapping unknown node IDs. Right?